hummingbot/hummingbot
# Feat/update okx connector

**Before submitting this PR, please make sure**:


**A description of the changes proposed in the pull request**:
This is a general update of the OKX connector to make it compliant with all latest standards for connectors.
The task also includes the following changes intended to improve the process to create new centralized connectors:
Refactoring of order book and user stream data source classes to move common code to the super class. The refactoring was applied to Binance, Kucoin, GateIO and OKX connectors.
The logic that was previously in the OrderBookDataSource class to get all trading pairs and the last traded prices is moved with this PR to the connector class. The change was also applied to Binance, Kucoin, GateIO and OKX connectors. All other existing connectors have been updated so that the connector class implements the required methods, but the logic that gets the trading pairs and the traded prices is still in their OrderBookDataSource classes.
Created a new abstract test class for exchange classes that includes all the basic tests. The intention is that new connectors will be able to create the exchange class tests by subclassing the abstract test class and implementing all abstract methods intended to return the mock responses from the server with the appropriate format.

**Tests performed by the developer**:
All unit tests are passing in green.
Tested Binance and OKX connectors running a Pure Market Making strategy.


**Test summary**:
- Connect Binance, Gate_io, and OKX ( use valid and invalid API keys) and confirm correct response is shown in the client.
- Check the connector table and confirmed that the right status is displayed, also run the balance command and successfully show each exchange's current balance.
- Create a PMM strategy - Tested on Binance, Gate_io, and OKX
        * Market auto complete - Binance: Ok
        * Market auto complete - Gate_io:
        * Market auto-complete - OKX: Ok
- Start the strategy and run status —live
- Order cancelation and creation: OK
          * Cancel one of the orders created by the bot in the exchange: Ok
          * Manually create an order in the exchange and cancel: Did not affect the bot, Ok
- Data integrity (best price, order book, and last price)
        * Binance: Ok
        * Gate_io: Ok
        * OKX: Ok
-  Filled event: Ok
-  Check history: OK
- Verify fees paid: Ok
- Test fast and long refresh rate: Ok
-  Test multiple levels: OK
-  Build a docker image and create a docker instance
- Connect Binance, Gate_io, and OKX API keys
- Run a basic PMM for each of the connectors (Binance, Gate_io, and OKX)

**Tips for QA testing**:
Test OKX connector.
Validate that `connect` command does not fail (when used to configure a connector and also when executed without parameters to check the general status of connectors).
Validate that `balance` command does not fail.


This PR comes from https://github.com/CoinAlpha/hummingbot/pull/211

PRP: https://snapshot.org/#/hbot-prp.eth/proposal/0x89c00291fbb4b705472d987198c5071e4e21ce3c54bf49fd784dd60a7060095f
